spring-kafka/src/test/java/org/springframework/kafka/listener/ConcurrentMessageListenerContainerTests.java

```diff
@@ -608,4 +614,116 @@ public void handle(Exception thrownException, ConsumerRecord<?, ?> record) {
+		// this consumer is positioned at 1, the next offset after the successfully
+		// processed 'foo'
+		// it has not been updated because 'bar' failed
+		assertThat(consumer.position(new TopicPartition(topic9, 0))).isEqualTo(1);
+		// this consumer is positioned at 1, the next offset after the successfully
+		// processed 'qux'
+		// it has been updated even 'baz' failed
+		assertThat(consumer.position(new TopicPartition(topic9, 1))).isEqualTo(2);
+		logger.info("Stop ack on error");
+	}
+
+	@Test
+	public void testRebalanceWithSlowConsumer() throws Exception {
+		this.logger.info("Start auto");
+		Map<String, Object> props = KafkaTestUtils.consumerProps("test101", "false", embeddedKafka);
+		props.put(ConsumerConfig.FETCH_MIN_BYTES_CONFIG, "20000");
+		DefaultKafkaConsumerFactory<Integer, String> cf = new DefaultKafkaConsumerFactory<Integer, String>(props);
+		ContainerProperties containerProps = new ContainerProperties(topic1);
+		ConcurrentMessageListenerContainer<Integer, String> container =
+				new ConcurrentMessageListenerContainer<>(cf, containerProps);
+		ConcurrentMessageListenerContainer<Integer, String> container2 =
+				new ConcurrentMessageListenerContainer<>(cf, containerProps);
+		final CountDownLatch latch = new CountDownLatch(8);
+		final Set<String> listenerThreadNames = Collections.synchronizedSet(new HashSet<String>());
+		List<String> receivedMessages = Collections.synchronizedList(new ArrayList<>());
+		containerProps.setMessageListener(new MessageListener<Integer, String>() {
+
+			@Override
+			public void onMessage(ConsumerRecord<Integer, String> message) {
+				System.out.println("auto: " + message + " on " + Thread.currentThread().getName());
+				listenerThreadNames.add(Thread.currentThread().getName());
+				try {
+					Thread.sleep(2000);
+				}
+				catch (InterruptedException e) {
+					// ignore
+				}
+				receivedMessages.add(message.value());
+				listenerThreadNames.add(Thread.currentThread().getName());
+				latch.countDown();
+			}
+		});
+		container.setConcurrency(1);
+		container2.setConcurrency(1);
+		container.setBeanName("testAuto");
+		container2.setBeanName("testAuto2");
+		container.start();
+		ContainerTestUtils.waitForAssignment(container, embeddedKafka.getPartitionsPerTopic());
+		Map<String, Object> senderProps = KafkaTestUtils.producerProps(embeddedKafka);
+		ProducerFactory<Integer, String> pf = new DefaultKafkaProducerFactory<Integer, String>(senderProps);
+		KafkaTemplate<Integer, String> template = new KafkaTemplate<>(pf);
+		template.setDefaultTopic(topic1);
+		template.sendDefault(0, 0, "foo");
+		template.sendDefault(0, 2, "bar");
+		template.sendDefault(0, 0, "baz");
+		template.sendDefault(0, 2, "qux");
+		template.sendDefault(1, 2, "corge");
+		template.sendDefault(1, 2, "grault");
+		template.sendDefault(1, 2, "garply");
+		template.sendDefault(1, 2, "waldo");
+		template.flush();
+		container2.start();
```

Comment:
Yeah, it needs to be stopped. What we want is to cause a rebalance. Probably we will need some additional way to prove that a container didn't process an entire batch before stopping.